Add initial init/term assignment

michalk8 opened the pull request:
**IMPORTANT: Please search among the [Pull requests](../pulls) before creating one.**

## Description
<!-- Clearly and concisely describe your changes. This section will be shown in the docs. -->
TODO.

## How has this been tested?
<!-- Describe in detail how you've tested your changes. -->
TODO.

## Closes
<!-- If applicable, type `closes #XXXX` in your comment to auto-close the issue that this PR fixes. -->
closes #908

Marius1311 commented:
Let me know once you want me to review/test this.

Marius1311 commented:
I'm trying to run this to a GPCCA estimator initialized from a VelocityKernel. I computed 4 marcostates and run `g.predict()´, which gives me:
```pytb
---------------------------------------------------------------------------
AttributeError                            Traceback (most recent call last)
Input In [9], in <cell line: 1>()
----> 1 g.predict()
      2 g.plot_terminal_states(discrete=True, legend_loc='right', s=100)

File ~/Projects/cellrank/cellrank/estimators/terminal_states/_gpcca.py:274, in GPCCA.predict(self, method, which, n_cells, alpha, stability_threshold, n_states)
    271     raise RuntimeError("Compute macrostates first as `.compute_macrostates()`.")
    273 # fmt: off
--> 274 if len(self._macrostates.cat.categories) == 1:
    275     logg.warning(f"Found only one macrostate. Making it the single {which} state")
    276     self.set_states_from_macrostates(names=None, which=which, n_cells=n_cells, params=self._create_params())

AttributeError: 'StatesHolder' object has no attribute 'cat'
```

Marius1311 commented:
Note: this wasn't actually done with a VelocityKernel, I saved the transition matrix to file and then worked with a precomputed kernel because of https://github.com/theislab/cellrank/issues/933

Marius1311 commented:
Hi @michalk8, can I try again?

michalk8 commented:
> Hi @michalk8, can I try again?

Yes, the above error should've been fixed, but most likely there are other bugs still.

Marius1311 commented:
Hi Mike, just going through your PR. This is beautiful work, as always. Of course, I do still have a few comments 🙃

My first comment concerns the two methods `g.plot_states` and `g.plot_macrostates`, for GPCCA estimator `g`. I'm just not sure we really need two different plotting functions here - essentially, they do the same thing and the behavior could be entirely controlled via the `which` parameter. I think we could only keep the `g.plot_macrostates` method and use
- `which=all` to show all macrostates, mimicking the current output of `g.plot_macrostates`
- `which=initial/terminal`, to show initial or terminal states, mimicking the current output of `g.plot_states` 
- `which=['state_1', 'state_2', ...]` to plot a custom list of macrostates

One reason for my suggesting this is that the two methods also behave a bit differently at the moment. For example, calling `g.plot_macrostates(discrete=True, legend_loc='right', s=100)` gives me Fig. 1, while `g.plot_states(which='terminal', legend_loc='right', size=100)` gives me Fig. 2 below. Notice how the size parameter is set to the same value, yet in Fig. 1, this seems to only have an effect on the highlighted cells in color, while in Fig. 2, it seems to affect all cells. I like the former behavior better, e.g. just increasing the size of the cells we are highlighting, and I even think this should be done by default.

Marius1311 commented:
My next comment concerns the `g.predict` method used to classify the macrostates. While `g.predict(which='terminal')` works fine and gives me the expected output, `g.predict(which='initial')` does not work because it gives me the exact same states. In this PR, we don't want to change how we used to compute initial states via outliers in the coarse-grained stationary distribution; we only want to change how we interact with these functionalities. Fig. 1 shows the output I get from running 
``` 
g.predict(which='initial')
g.plot_states(which='initial', legend_loc='right', s=100)
```
So clearly these are the terminal states. Interestingly, only the highlighted cells are colored as in the macrostates plotting function, see my comment above. 

The ways in which we identify initial and terminal states differ, and depending on the `which` parameter, we have to change method defaults. I think this gets leads to confusion when we do it via a single method, and I would suggest having two methods:
- `g.predict_terminal_states`: pretty much with the current parameters and defaults of `g.predict(which='terminal`)
- `g.predict_initial_states`: similar to what we used to have in `g._compute_initial_states()`; which seems to have been removed in this PR. 

## Fig. 1
<img width="549" alt="Screenshot 2022-12-18 at 13 16 27" src="https://user-images.githubusercontent.com/24494650/208297885-26501a5a-c8fe-425b-a89e-2978e0992541.png">

Marius1311 commented:
Before classifying macrostates as initial or terminal, the estimator prints an empty field for terminal states, but not for initial. I would suggest to make this consistent, either show for both initial and terminal or for none. 

<img width="1076" alt="Screenshot 2022-12-18 at 14 44 12" src="https://user-images.githubusercontent.com/24494650/208301698-45d31a6f-2857-41ba-bf88-88126606e3e2.png">

Marius1311 commented:
I would strongly vote for removing the `g.compute_states()` alisas, it just adds confusion. While this method sounds super similar to `g.compute_macrostates`, it does something really different.

Marius1311 commented:
> Before classifying macrostates as initial or terminal, the estimator prints an empty field for terminal states, but not for initial. I would suggest making this consistent, either show for both initial and terminal or for neither.
> 
> <img alt="Screenshot 2022-12-18 at 14 44 12" width="1076" src="https://user-images.githubusercontent.com/24494650/208301698-45d31a6f-2857-41ba-bf88-88126606e3e2.png">

Even once I predict initial states, this still only shows terminal states. I would add initial states to this representation - we should treat both equally.

Marius1311 commented:
I would also vote for writing macrostates to the underlying AnnData's `.obs` field, same as for initial and terminal states. I have a usecase in which I want to visualize the distribution of marker genes over macrostates - I can manually write the macrostates to AnnData but then I still have the wrong color assignment, it would be much more convenient to have this written automatically upon computation.

Marius1311 commented:
To get a feeling for how I plan to use this PR in the new initial & terminal states tutorial, please check this related PR in the notebooks repo: https://github.com/theislab/cellrank_notebooks/pull/56

Marius1311 commented:
Discussion outcome
- remove alias `g.compute_states`
- unify `g.plot_states`and `g.plot_macrostates`
- introduce `g.predict_terminal_states`and `g.predict_initial_states`, where the latter uses the coarse-grained stationary distribution (check old code)
- no overlap between initial and terminal states (Warning and rename, or throw an error)
- only print initial and terminal states (not macrostates)
- writing macrostates to AnnData

Marius1311 commented:
Hi Mike, happy new year! Did you have time to implement some of the changes already?

Marius1311 commented:
Thanks @michalk8, here my comments:
- As suggested before, I would prefer to have a `g.plot_macrostates`, rather than `g.plot_states` method. I think this makes more sense, as everything we are dealing with are macrostates (initial macrostates, terminal macrostates, etc. ). Is there a reason you choose to call the method `g.plot_states`? It's quite confusing now, we have a `g.compute_macrostates` method, but then a `g.plot_states` method. Also, internally, we refer to stuff as `macrostates`. 
- Plotting does not work yet, calling `g.compute_macrostates(cluster_key='clusters', n_states=[4, 12], method='brandts')`followed by `g.plot_states()` gives Error 1 below
- `g.plot_states` does not have a `which` parameter as we discussed above

I think it makes more sense to have another meeting to discuss the anticipated workflow before we continue. 

## Error 1
```
---------------------------------------------------------------------------
TypeError                                 Traceback (most recent call last)
Cell In [19], line 1
----> 1 g.plot_states()

File ~/Projects/cellrank/cellrank/estimators/mixins/_utils.py:445, in register_plotter.<locals>.wrapper(wrapped, instance, args, kwargs)
    442 else:
    443     raise TypeError(f"Unable to plot a value of type `{type(obj)}`.")
--> 445 return wrapped(
    446     *args,
    447     _data=data,
    448     _colors=colors,
    449     _title=f"{attr} states",
    450     discrete=discrete,
    451     **kwargs,
    452 )

File ~/miniforge3/envs/py39_arm/lib/python3.9/site-packages/scanpy/_utils/__init__.py:119, in deprecated_arg_names.<locals>.decorator.<locals>.func_wrapper(*args, **kwargs)
    117 # reset filter
    118 warnings.simplefilter('default', DeprecationWarning)
--> 119 return func(*args, **kwargs)

File ~/Projects/cellrank/cellrank/estimators/mixins/_utils.py:366, in _plot_dispatcher(self, states, color, discrete, mode, time_key, same_plot, title, cmap, **kwargs)
    336 """
    337 Plot continuous or categorical observations in an embedding or along pseudotime.
    338 
   (...)
    363 %(just_plots)s
    364 """
    365 if discrete:
--> 366     return _plot_discrete(
    367         self,
    368         states=states,
    369         color=color,
    370         same_plot=same_plot,
    371         title=title,
    372         cmap=cmap,
    373         **kwargs,
    374     )
    376 return _plot_continuous(
    377     self,
    378     states=states,
   (...)
    385     **kwargs,
    386 )

File ~/Projects/cellrank/cellrank/estimators/mixins/_utils.py:166, in _plot_discrete(self, _data, _colors, _title, states, color, title, same_plot, cmap, **kwargs)
    153 def _plot_discrete(
    154     self: PlotterProtocol,
    155     _data: pd.Series,
   (...)
    163     **kwargs: Any,
    164 ) -> None:
    165     if not isinstance(_data, pd.Series):
--> 166         raise TypeError(
    167             f"Expected `data` to be of type `pandas.Series`, found `{type(_data).__name__}`."
    168         )
    169     if not is_categorical_dtype(_data):
    170         raise TypeError(
    171             f"Expected `data` to be `categorical`, found `{infer_dtype(_data)}`."
    172         )

TypeError: Expected `data` to be of type `pandas.Series`, found `NoneType`.
```

michalk8 commented:
> Thanks @michalk8, here my comments:
> 
>     * As suggested before, I would prefer to have a `g.plot_macrostates`, rather than `g.plot_states` method. I think this makes more sense, as everything we are dealing with are macrostates (initial macrostates, terminal macrostates, etc. ). Is there a reason you choose to call the method `g.plot_states`? It's quite confusing now, we have a `g.compute_macrostates` method, but then a `g.plot_states` method. Also, internally, we refer to stuff as `macrostates`.
> 
>     * Plotting does not work yet, calling `g.compute_macrostates(cluster_key='clusters', n_states=[4, 12], method='brandts')`followed by `g.plot_states()` gives Error 1 below
> 
>     * `g.plot_states` does not have a `which` parameter as we discussed above
> 
> 
> I think it makes more sense to have another meeting to discuss the anticipated workflow before we continue.
> ## Error 1
> 
> ```
> ---------------------------------------------------------------------------
> TypeError                                 Traceback (most recent call last)
> Cell In [19], line 1
> ----> 1 g.plot_states()
> 
> File ~/Projects/cellrank/cellrank/estimators/mixins/_utils.py:445, in register_plotter.<locals>.wrapper(wrapped, instance, args, kwargs)
>     442 else:
>     443     raise TypeError(f"Unable to plot a value of type `{type(obj)}`.")
> --> 445 return wrapped(
>     446     *args,
>     447     _data=data,
>     448     _colors=colors,
>     449     _title=f"{attr} states",
>     450     discrete=discrete,
>     451     **kwargs,
>     452 )
> 
> File ~/miniforge3/envs/py39_arm/lib/python3.9/site-packages/scanpy/_utils/__init__.py:119, in deprecated_arg_names.<locals>.decorator.<locals>.func_wrapper(*args, **kwargs)
>     117 # reset filter
>     118 warnings.simplefilter('default', DeprecationWarning)
> --> 119 return func(*args, **kwargs)
> 
> File ~/Projects/cellrank/cellrank/estimators/mixins/_utils.py:366, in _plot_dispatcher(self, states, color, discrete, mode, time_key, same_plot, title, cmap, **kwargs)
>     336 """
>     337 Plot continuous or categorical observations in an embedding or along pseudotime.
>     338 
>    (...)
>     363 %(just_plots)s
>     364 """
>     365 if discrete:
> --> 366     return _plot_discrete(
>     367         self,
>     368         states=states,
>     369         color=color,
>     370         same_plot=same_plot,
>     371         title=title,
>     372         cmap=cmap,
>     373         **kwargs,
>     374     )
>     376 return _plot_continuous(
>     377     self,
>     378     states=states,
>    (...)
>     385     **kwargs,
>     386 )
> 
> File ~/Projects/cellrank/cellrank/estimators/mixins/_utils.py:166, in _plot_discrete(self, _data, _colors, _title, states, color, title, same_plot, cmap, **kwargs)
>     153 def _plot_discrete(
>     154     self: PlotterProtocol,
>     155     _data: pd.Series,
>    (...)
>     163     **kwargs: Any,
>     164 ) -> None:
>     165     if not isinstance(_data, pd.Series):
> --> 166         raise TypeError(
>     167             f"Expected `data` to be of type `pandas.Series`, found `{type(_data).__name__}`."
>     168         )
>     169     if not is_categorical_dtype(_data):
>     170         raise TypeError(
>     171             f"Expected `data` to be `categorical`, found `{infer_dtype(_data)}`."
>     172         )
> 
> TypeError: Expected `data` to be of type `pandas.Series`, found `NoneType`.
> ```

Updated to reflect all the changes as mentioned above. However, I removed the default `which`, because I think user should always supply it.

Marius1311 commented:
Oh, exciting, thanks for the updates! I will check this out soon! Looking fwd. to this.

Marius1311 commented:
Thanks, Mike, here are a number of comments (we're alomst there!)
- in `g.plot_macrostates`, please change `which="macro"` to `which="all"` to show all macrostates. I think this makes more sense syntactically - you're calling a method that has "macrostates" in its name, so I don't think we should have a parameter that re-iterates this. 
- Please remove the `g.predict()` alias for the `g.predict_terminal_states()`. It is unintuitive why that alias points to  `g.predict_terminal_states()` and not  `g.predict_initial_states()` as we want to treat them equally in CR2

I'll post these now and add a few more later

Marius1311 commented:
There is a bit of an inconsistency now between automatic and manual identification of initial and terminal states:
- In automatic mode, we have two methods, `g.predict_terminal_states` and `g.predict_initial_states`
- In manual mode, we only have one method, `g.set_states_from_macrostates`, and we control the behavior via the `which` parameter

You could argue that the second mode of action is a bit similar to macrostate plotting. However, when we plot macrostates, the method is called `g.plot_macrostates`, which I much prefer over the old `g.plot_states`.  In general, I would avoid the term `states` and restrict our vocabulary to `macrostates`, `terminal_states`and `initial_states`, to keep things simple. 

So I suggest having `g.set_terminal_states_from_macrostates`and `g.set_initial_states_from_macrostates` for consistency. What is your opinion here @michalk8 ?

michalk8 commented:
> in g.plot_macrostates, please change which="macro" to which="all" to show all macrostates. I think this makes more sense syntactically - you're calling a method that has "macrostates" in its name, so I don't think we should have a parameter that re-iterates this.

Ok!


> Please remove the g.predict() alias for the 

`predict` cannot be removed since it's an abstract method of the base estimator (alongside `fit`).

> So I suggest having g.set_terminal_states_from_macrostatesand g.set_initial_states_from_macrostates for consistency. What is your opinion here @michalk8 ?

I think the proposed names are too long.

Marius1311 commented:
> > in g.plot_macrostates, please change which="macro" to which="all" to show all macrostates. I think this makes more sense syntactically - you're calling a method that has "macrostates" in its name, so I don't think we should have a parameter that re-iterates this.
> 
> Ok!
> 
> > Please remove the g.predict() alias for the
> 
> `predict` cannot be removed since it's an abstract method of the base estimator (alongside `fit`).

That's an issue! It's very arbitrary what predict points to! We should talk about whether we really want our estimators to follow this fit-predict mechanism - I have the feeling this is a bit forced at the moment and might not actually work for us. 


> 
> > So I suggest having g.set_terminal_states_from_macrostatesand g.set_initial_states_from_macrostates for consistency. What is your opinion here @michalk8 ?
> 
> I think the proposed names are too long.

Marius1311 commented:
Note to self: the coarse-grained stationary distribution changes compared to the tutorial version on master, I think this has something to do with the way in which I define the VelocityKernel

## Previous version of the velocity kernel

12 macrostates 
We correctly identify the Ngn3 low EP 1 state as initial: 
<img width="311" alt="Screenshot 2023-02-23 at 14 45 13" src="https://user-images.githubusercontent.com/24494650/220924872-4fa44f0e-2275-4845-84fd-53c202b5e2fb.png">

15 macrostates
We correctly identify the Ngn3 low EP 1 state as initial: 
<img width="311" alt="Screenshot 2023-02-23 at 14 46 18" src="https://user-images.githubusercontent.com/24494650/220925128-345e1aa4-b4e2-4060-a290-0465157c0591.png">

Marius1311 commented:
Ok, in summary: 
- We agreed on the first point about "macro" -> "all" (and I think this would be a good default value)
- We can further discuss the `predict` alias tomorrow, as well as the naming/mechanism for semi-manual macrostate finding
- One more feature that would be fantastic to add to the `plot_macrostates` method is a mode `which="initial_and_terminal"`, which shows both initial and terminal states. To tell them apart, the names should be modified in the legend such that they show up as `initial: macrostate_1, terminal: macrostate_2` etc. I suggested this above already, what do you think? I think this would be a very useful aggregate plotting mode.

Marius1311 commented:
Discussion outcome:
- Unify `set_states_from_macrostates` with the `set_states` method into two methods: `set_initial_states` and `set_terminal_states`, which can either set terminal/initial states from macrostates or purely manually from pandas Series or similar, depending on the input.

Marius1311 commented:
@WeilerP, this PR introduces quite a few changes to the way we compute initial and terminal states - not the actual methods we use under the hood, only how we call the corresponding methods etc. Just to let you know that there will be some breaking changes here. I can guide you through this in more detail next week.

Marius1311 commented:
Uh, I'm excited to check this out @michalk8!
